### PR TITLE
Added exponential placer cost function

### PR DESCRIPTION
**Pull request by jack-melchert**
Some very congested applications have trouble routing using the default placer hpwl cost function. This branch allows the user to set the environmental variable PNR_PLACER_EXP, which will change the cost function to exp(hpwl, PNR_PLACER_EXP). In my tests, setting this value to a value greater than 1 significantly reduces the routing issues in congested apps.

### CODE REVIEW

File: thunder/src/util.cc [hunk 1]
```diff
@@ -44,7 +44,12 @@ double get_hpwl(const ::vector<Net> &netlist,
             if (pos.y > ymax)
                 ymax = pos.y;
         }
-        hpwl += (xmax - xmin) + (ymax - ymin);
+        int exp = 1;
+        char* dist_exp = std::getenv("PNR_PLACER_EXP");
```
Comment (Kuree) [inline thunder/src/util.cc]: Can you refactor this such that the option is set in the placer? This function gets called very frequently and getting env is not efficient.

**Review by Kuree — CHANGES_REQUESTED** [review]

File: thunder/src/util.cc (hunk 1 above)
Comment (Kuree) [reply thunder/src/util.cc]: For instance, to make it easier to implement, you can have a static boolean variable that gets set in the placer, but read out here.

Comment (jack-melchert) [reply thunder/src/util.cc]: Does this look better?

File: thunder/src/detailed.cc [hunk 2]
```diff
@@ -277,6 +279,15 @@ void DetailedPlacer::set_fold_reg(const ::vector<::string> &cluster_blocks,
     }
 }
 
+void DetailedPlacer::set_placer_cost_exp() {
+    int exp = 1;
+    char* dist_exp = std::getenv("PNR_PLACER_EXP");
+    if (dist_exp != NULL) {
```
Comment (Kuree) [inline thunder/src/detailed.cc]: Use either

```C++
if (!dist_exp)
```

of
```C++
if (dist_exp != nullptr)
```

`NULL` is old C syntax which is not recommended in C++.

Other than this, the PR looks good to me.

**Review by Kuree — CHANGES_REQUESTED** [review]

**Review by Kuree — APPROVED** [review]

**Comment by jack-melchert** [comment]
Is the Mac wheel test stuck? Its been running for over an hour now.

**Comment by Kuree** [comment]
It seems so. I don't have access to mac so it's hard for me to debug.

**Comment by Kuree** [comment]
Added in [116a52b](https://github.com/Kuree/cgra_pnr/commit/116a52b0f61a474e6d7594baf0b7f358fb1285e5)

**Pull request closed without merging**